### PR TITLE
[Linker] Propagate `nobuiltin` attributes when linking known libcalls

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
As discussed in https://discourse.llvm.org/t/rfc-libc-ffreestanding-fno-builtin.

LLVM ascribes special semantics to several functions that are known to
be `libcalls`. These are functions that middle-end optimizations may
transforms calls into or perform optimizations based off of known
semantics. However, these assumptions require an opaque function call to
be known valid. In situations like LTO or IR linking it is possible to
bring a libcall definition into the current module. Once this happens,
we can no longer make any guarantees about the semantics of these
functions.

We currently attempt to solve this by preventing all inlining if the
called function has `no-builtin` https://reviews.llvm.org/D74162.
However, this is overly pessimistic as it prevents all inlining even for
non-libcall functions.

This patch modifies the IRMover class to track known libcalls enabled
for the given target. If we encounter a known libcall during IR linking,
we then need to append the `nobuiltin` attribute to the destination
module. Afterwards, all new definitions we link in will be applied as
well.

### CODE REVIEW

File: llvm/lib/Linker/IRMover.cpp
Comment (arsenm) [inline llvm/lib/Linker/IRMover.cpp:1631]:
```suggestion
          F->getContext(), Attribute::NoBuiltin));
```

File: llvm/include/llvm/Linker/IRMover.h
Comment (arsenm) [inline llvm/include/llvm/Linker/IRMover.h:83]: I'd expect this to be never null

File: llvm/lib/Linker/IRMover.cpp
Comment (jhuber6) [reply llvm/lib/Linker/IRMover.cpp]: Updated

File: llvm/include/llvm/Linker/IRMover.h
Comment (jdoerfert) [inline llvm/include/llvm/Linker/IRMover.h:83]:
```suggestion
    bool checkLibcalls(GlobalValue &GV) {
```

File: llvm/include/llvm/Linker/IRMover.h
Comment (jdoerfert) [inline llvm/include/llvm/Linker/IRMover.h:81]:
```suggestion
    void updateLibcalls(const Triple &Triple);
```

File: llvm/lib/Linker/IRMover.cpp
Comment (jdoerfert) [inline llvm/lib/Linker/IRMover.cpp:1631]: Don't you have to do this anyway at the end? Just have a single location to do it.

File: llvm/lib/Linker/IRMover.cpp
Comment (jdoerfert) [inline llvm/lib/Linker/IRMover.cpp:1787]:
```suggestion
void IRMover::LibcallHandler::updateLibcalls(const Triple &Triple) {
```

**Review by jdoerfert — CHANGES_REQUESTED** [review]
I don't think this works without a module flag because the user might switch the order; linking the application into libc, or linking something else later into the application with libc. Either way, we wouldn't link in libcalls but we already had them. Still we can't perform libcall reasoning on the new functions.

**Comment by jhuber6** [comment]
> I don't think this works without a module flag because the user might switch the order; linking the application into libc, or linking something else later into the application with libc. Either way, we wouldn't link in libcalls but we already had them. Still we can't perform libcall reasoning on the new functions.

Doesn't  this cover that case? The test file links the application and `libc` definition in either order. Both cases put `nobuiltin` on all the resulting functions.

File: llvm/lib/Linker/IRMover.cpp
Comment (jhuber6) [reply llvm/lib/Linker/IRMover.cpp:1787]: Doesn't work because the variable is named `Triple`. I could rename it to `TheTriple` if you'd prefer.

File: llvm/lib/Linker/IRMover.cpp
Comment (jhuber6) [reply llvm/lib/Linker/IRMover.cpp:1631]: No, there are two cases.

1. Linking in a module that contains libcalls.

When this happens we need to modify *all* definitions in the destination / merged module. However, once we do this once we know that the merged module is already processed.

2. Linking a module into a module that already contains libcalls

Here we only need to add the attributes to the definitions as they come in because we know the destination module has already been handled (If the boolean is set true). I did it this way to avoid looping over every single function that we've already touched. I could make it do that if you think the code clarity is better.

**Comment by jhuber6** [comment]
I wonder if this should add `"no-builtins"` as well. I don't really know what that does that `"no-builtins"` doesnt, but I noticed that there's some logic in the `TargetLibraryInfo` that optionally takes a function and disables all libcalls if it has `"no-builtins"` so it's possible something we should propagate.

**Comment by jhuber6** [comment]
ping

**Comment by jhuber6** [comment]
ping

**Comment by gchatelet** [comment]
@jhuber6 I don't know the linking part enough to do a good review here so I will leave it to more knowledgeable people. @teresajohnson would you be able to help?

File: llvm/test/Linker/libcalls.ll
Comment (jdoerfert) [inline llvm/test/Linker/libcalls.ll:8]: Add another function here, so the order in the output should be
something
strlen
something
and all should be annotated.

**Review by jdoerfert — APPROVED** [review]
The test seems to show my concerns were unwarranted.
See the comment there to make it a little more elaborate.
The code looks good and I don't think it'll impact compile time.

**Pull request merged**
